### PR TITLE
EN-52985: Custom index

### PR DESCRIPTION
**Pull request by chitang**

### CODE REVIEW

**Comment by chitang** [comment]
This PR and the one in DC are kinda large.  But they are almost 100% boilerplate copied from Rollup or IndexDirective.

File: soda-fountain-lib/src/main/scala/com/socrata/soda/clients/datacoordinator/HttpDataCoordinatorClient.scala [hunk 1]
```diff
@@ -763,6 +766,25 @@ abstract class HttpDataCoordinatorClient extends DataCoordinatorClient {
...
+        val request = indexReq(host, dataset).get
+        httpClient.execute(request).run { r =>
+          errorFrom(r) match {
+            case None => r.value[Seq[IndexInfo]]() match {
+              case Right(resp) => IndexResult(resp)
+              case Left(err) => throw new Exception("Unable to parse response from data coordinator: " + err.english)
+            }
+            case Some(NoSuchDataset(dataset)) => DatasetNotFoundResult(dataset)
+            case Some(err) =>
+              throw new Exception(s"Unexpected error from data-coordinator getting rollups for dataset $dataset: $err")
```
Comment (catstavi) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/clients/datacoordinator/HttpDataCoordinatorClient.scala:781]: * getting indexes? (not rollups)

File: soda-fountain-lib/src/main/scala/com/socrata/soda/clients/datacoordinator/IndexMutation.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import com.rojoma.json.v3.util.JsonUtil
+import com.socrata.soda.server.id.IndexName
+
+sealed abstract class IndexMutation extends DataCoordinatorInstruction {
+  override def toString = JsonUtil.renderJson(asJson)
+}
+
+case class CreateOrUpdateIndexInstruction(name: IndexName, expressions: String, filter: Option[String]) extends IndexMutation {
+  def asJson = JObject(Map(
+    "c"     -> JString("create or update index"),
```
Comment (catstavi) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/clients/datacoordinator/IndexMutation.scala:13]: I see we use "c" as some type of standard key in lots of different DataCoordinatorInstruction things (and I don't have a problem with it), just wondering what the deal is/what it stands for?

File: soda-fountain-lib/src/main/scala/com/socrata/soda/server/highlevel/DatasetDAOImpl.scala [hunk 3]
```diff
@@ -507,6 +507,95 @@ class DatasetDAOImpl(dc: DataCoordinatorClient,
...
+              val instruction = CreateOrUpdateIndexInstruction(name, expressions, spec.filter)
+              dc.update(datasetRecord.handle, datasetRecord.schemaHash, expectedDataVersion, user,
+                Iterator.single(instruction)) {
+                case DataCoordinatorClient.NonCreateScriptResult(report, etag, copyNumber, newVersion, newShapeVersion, lastModified) =>
+                  store.updateVersionInfo(datasetRecord.systemId, newVersion, lastModified, None, copyNumber, None)
+                  IndexCreatedOrUpdated
+                case DataCoordinatorClient.NoSuchIndexResult(_, _) =>
+                  IndexNotFound(name)
+                case DataCoordinatorClient.InvalidIndexResult(name, _) =>
+                  RollupError(name.name)
```
Comment (catstavi) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/server/highlevel/DatasetDAOImpl.scala:530]: RollupError?

**Comment by catstavi** [comment]
(those are all my comments/questions!)

File: soda-fountain-lib/src/main/scala/com/socrata/soda/clients/datacoordinator/HttpDataCoordinatorClient.scala (hunk 1 above)
Comment (chitang) [reply soda-fountain-lib/src/main/scala/com/socrata/soda/clients/datacoordinator/HttpDataCoordinatorClient.scala:781]: yes

File: soda-fountain-lib/src/main/scala/com/socrata/soda/clients/datacoordinator/IndexMutation.scala (hunk 2 above)
Comment (chitang) [reply soda-fountain-lib/src/main/scala/com/socrata/soda/clients/datacoordinator/IndexMutation.scala:13]: It stands for command.  Defined in
 private def decodeCommand(index: Long, json: JValue): Command =
      withObjectFields(index, json) { accessor =>
        import accessor._

        get[String]("c") match {

File: soda-fountain-lib/src/main/scala/com/socrata/soda/server/highlevel/DatasetDAOImpl.scala (hunk 3 above)
Comment (chitang) [reply soda-fountain-lib/src/main/scala/com/socrata/soda/server/highlevel/DatasetDAOImpl.scala:530]: Obviously, i copied from rollup :)

**Review by catstavi — APPROVED** [review]

**Pull request merged**